libensemble/generators.py

```diff
@@ -0,0 +1,227 @@
+
+    def ask(self, num_points: int = 0, *args, **kwargs) -> (Iterable, Optional[npt.NDArray]):
+        _, self.last_ask = self.outbox.get()
+        return self.last_ask["calc_out"]
+
+    def tell(self, results: npt.NDArray, tag: int = EVAL_GEN_TAG) -> None:
+        if results is not None:
+            results = self._set_sim_ended(results)
+            self.inbox.put((tag, {"libE_info": {"H_rows": results["sim_id"], "persistent": True, "executor": None}}))
+        else:
+            self.inbox.put((tag, None))
+        self.inbox.put((0, results))
+
+    def final_tell(self, results: npt.NDArray) -> (npt.NDArray, dict, int):
+        self.tell(results, PERSIS_STOP)
+        return self.gen.result()
+
+
+class APOSMM(LibEnsembleGenInterfacer):
+    def __init__(
+        self, gen_specs: dict, History: npt.NDArray = [], persis_info: dict = {}, libE_info: dict = {}
+    ) -> None:
+        from libensemble.gen_funcs.persistent_aposmm import aposmm
+
+        gen_specs["gen_f"] = aposmm
+        if not persis_info:
+            persis_info = add_unique_random_streams({}, 4)[1]
+            persis_info["nworkers"] = 4
+        super().__init__(gen_specs, History, persis_info, libE_info)
+
+    def initial_ask(self, num_points: int = 0, *args) -> npt.NDArray:
+        return super().initial_ask(num_points, args)[0]
+
+    def ask(self, num_points: int = 0) -> (npt.NDArray, npt.NDArray):
+        results = super().ask(num_points)
+        if any(results["local_min"]):
+            minima = results[results["local_min"]]
+            results = results[~results["local_min"]]
+            return results, minima
+        return results, np.empty(0, dtype=self.gen_specs["out"])
+
+    def tell(self, results: npt.NDArray, tag: int = EVAL_GEN_TAG) -> None:
+        super().tell(results, tag)
+
+    def final_tell(self, results: npt.NDArray) -> (npt.NDArray, dict, int):
+        return super().final_tell(results)
+
+
+class Surmise(LibEnsembleGenInterfacer):
+    def __init__(
+        self, gen_specs: dict, History: npt.NDArray = [], persis_info: dict = {}, libE_info: dict = {}
+    ) -> None:
+        from libensemble.gen_funcs.persistent_surmise_calib import surmise_calib
+
+        gen_specs["gen_f"] = surmise_calib
+        if ("sim_id", int) not in gen_specs["out"]:
+            gen_specs["out"].append(("sim_id", int))
+        super().__init__(gen_specs, History, persis_info, libE_info)
+        self.sim_id_index = 0
+
+    def _add_sim_ids(self, array: npt.NDArray) -> npt.NDArray:
+        array["sim_id"] = np.arange(self.sim_id_index, self.sim_id_index + len(array))
+        self.sim_id_index += len(array)
+        return array
+
+    def initial_ask(self, num_points: int = 0, *args) -> npt.NDArray:
+        return super().initial_ask(num_points, args)[0]
+
+    def ask(self, num_points: int = 0) -> (npt.NDArray, Optional[npt.NDArray]):
+        _, self.last_ask = self.outbox.get()
+        points = self._add_sim_ids(self.last_ask["calc_out"])
```

Comment:
Not clear to me why expect points then cancels. In Surmise gen_f main loop, there is a send_recv which will send points and then wait (for a tell). The next comm is a cancel (if there is one).